### PR TITLE
Implement in order aggregation (optimize_aggregation_in_order) for projections for tables with fully materialized projections

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement in order aggregation (`optimize_aggregation_in_order`) for fully materialized projections.

Cc: @amosbird @KochetovNicolai

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -5474,8 +5474,9 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
         // If optimize_aggregation_in_order = true, we need additional information to transform the projection's pipeline.
         auto attach_aggregation_in_order_info = [&]()
         {
-            for (const auto & key : keys)
+            for (const auto & desc : select.getQueryAnalyzer()->aggregationKeys())
```
Comment (amosbird) [inline src/Storages/MergeTree/MergeTreeData.cpp:5477]: What's the purpose of this change?

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -338,30 +392,70 @@ QueryPlanPtr MergeTreeDataSelectExecutor::read(
...
+                auto merge_threads = num_streams;
+                auto temporary_data_merge_threads = settings.aggregation_memory_efficient_merge_threads
+                    ? static_cast<size_t>(settings.aggregation_memory_efficient_merge_threads)
+                    : static_cast<size_t>(settings.max_threads);
 
-        if (!projection_pipe.empty())
-            build_aggregate_pipe(projection_pipe, true);
-        if (!ordinary_pipe.empty())
-            build_aggregate_pipe(ordinary_pipe, false);
+                InputOrderInfoPtr group_by_info = query_info.projection->input_order_info;
```
Comment (amosbird) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:416]: The code duplication is significant. If the `in_order_aggregate` optimization can only be applicated to one pipe, and if here `group_by_info` is the enabler, let's combine the two branches and only condition on this.

File: tests/queries/0_stateless/02320_projections_optimize_aggregation_in_order.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
```
Comment (amosbird) [inline tests/queries/0_stateless/02320_projections_optimize_aggregation_in_order.sh:1]: Could you rename the test to begin with `01710_`?

**Review by amosbird — CHANGES_REQUESTED** [review]

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp (hunk 2 above)
Comment (azat) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:416]: In one case we need to add transform, in another step, that's why there are two different branches.
I will reduce copy paste a little though, and I'm open to any suggestions.

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/MergeTree/MergeTreeData.cpp:5477]: Order should be preserved for in order aggregation (this has been covered in tests).

See https://github.com/ClickHouse/ClickHouse/pull/37469/commits/a966cfc682b41ec6fe8153bd35f61ab9d4cf7567.

File: tests/queries/0_stateless/02320_projections_optimize_aggregation_in_order.sh (hunk 3 above)
Comment (azat) [reply tests/queries/0_stateless/02320_projections_optimize_aggregation_in_order.sh:1]: OK (and I will add some clarification to existing `01710_projection_aggregation_in_order`)

**Comment by azat** [comment]
@KochetovNicolai can you please take a look? (or maybe someone else could/want?)

**Comment by azat** [comment]
>[Stress test (undefined, actions) — Hung check failed](https://s3.amazonaws.com/clickhouse-test-reports/37469/2b513acb2b8d703e737a700a9ccd5f7f323b605f/stateless_tests__aarch64__actions_.html)

```
        "elapsed": 2486.29842776,
        "is_cancelled": 1,
        "is_all_data_sent": 0,
        "read_rows": "500000",
        "read_bytes": "4000000",
        "total_rows_approx": "0",
        "written_rows": "500000",
        "written_bytes": "4000000",
        "memory_usage": "6427070",
        "peak_memory_usage": "6427070",
        "query": "INSERT INTO a FORMAT TSV",
```

```
$ pigz -cd clickhouse-server.stress.log.gz | grep -a -e '{11}' 
2022.05.31 19:00:26.118816 [ 2481 ] {11} <Debug> executeQuery: (from [::1]:60892) (comment: 01019_parallel_parsing_cancel.sh) INSERT INTO a FORMAT TSV (stage: Complete)
2022.05.31 19:00:26.163092 [ 2481 ] {11} <Trace> ContextAccess (default): Access granted: INSERT(x) ON test_15.a
2022.05.31 19:00:26.187395 [ 2481 ] {11} <Trace> ContextAccess (default): Access granted: SELECT(x) ON test_15.a
2022.05.31 19:00:26.203316 [ 2481 ] {11} <Trace> ContextAccess (default): Access granted: SELECT(x) ON test_15.a
2022.05.31 19:00:26.235590 [ 2481 ] {11} <Trace> ContextAccess (default): Access granted: SELECT(x) ON test_15.a
2022.05.31 19:00:26.447828 [ 2481 ] {11} <Trace> ContextAccess (default): Access granted: SELECT(x) ON test_15.a
2022.05.31 19:00:26.448177 [ 2481 ] {11} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
```

```
#19 0x0000000025199eba in DB::ExecutingInnerQueryFromViewTransform::onGenerate (this=0x7f19547dc428) at ../src/Processors/Transforms/buildPushingToViewsChain.cpp:564
```

- #37892

**Comment by azat** [comment]
Rebased (to fix conflicts with #37543)

File: src/Interpreters/Aggregator.cpp [hunk 4]
```diff
@@ -2631,45 +2685,75 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
...
-    for (size_t i = 0; i < params.aggregates_size; ++i)
-    {
-        const auto & aggregate_column_name = params.aggregates[i].column_name;
-        aggregate_columns[i] = &typeid_cast<const ColumnAggregateFunction &>(*block.getByName(aggregate_column_name).column).getData();
-    }
+    AggregateColumnsConstData aggregate_columns = params.makeAggregateColumnsData(block);
+    mergeWithoutKeyStreamsImpl(result, 0, block.rows(), aggregate_columns);
 
+    /// Early release memory.
+    block.clear();
```
Comment (amosbird) [inline src/Interpreters/Aggregator.cpp:2749]: If we always want to release memory here, it's better to change the signature to accept `Block` by value and use `std::move` to pass it. This explicitly express resource consumption.

**Review by amosbird — COMMENTED** [review]
LGTM, only one suggestion. @KochetovNicolai Could you help take a look?

**Review by KochetovNicolai — APPROVED** [review]
Looks ok

File: src/Interpreters/Aggregator.cpp (hunk 4 above)
Comment (azat) [reply src/Interpreters/Aggregator.cpp:2749]: Good idea, done.

File: src/Interpreters/Aggregator.cpp [hunk 5]
```diff
@@ -2631,45 +2685,69 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
...
     }
-
-    /// Early release memory.
-    block.clear();
 }
 
 template <typename Method, typename Table>
 void NO_INLINE Aggregator::mergeStreamsImpl(
-    Block & block,
+    Block && block,
```
Comment (amosbird) [inline src/Interpreters/Aggregator.cpp:2697]: This will not work as expect. It's still pass-by-reference (r-value) and it will not be destructed after function ends.

File: src/Interpreters/Aggregator.cpp [hunk 6]
```diff
@@ -2631,45 +2685,69 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
...
     else
-        mergeStreamsImplCase<true>(block, aggregates_pool, method, data, overflow_row);
+        mergeStreamsImplCase<true>(aggregates_pool, method, data, overflow_row, row_begin, row_end, aggregate_columns_data, key_columns);
 }
 
 
-void NO_INLINE Aggregator::mergeWithoutKeyStreamsImpl(
-    Block & block,
+void NO_INLINE Aggregator::mergeBlockWithoutKeyStreamsImpl(
+    Block && block,
```
Comment (amosbird) [inline src/Interpreters/Aggregator.cpp:2739]: same here

File: src/Interpreters/Aggregator.cpp (hunk 5 above)
Comment (azat) [reply src/Interpreters/Aggregator.cpp]: Indeed, thanks, fixed.

**Comment by azat** [comment]
>[Stateless tests (thread, actions) [1/3] — Some tests restarted, fail: 1, passed: 1397, skipped: 11](https://s3.amazonaws.com/clickhouse-test-reports/37469/0d4f78639e94f433e98a580d2c97a61a4193e962/stateless_tests__thread__actions__[1/3].html)

- `01183_custom_separated_format_http`

```
2022-06-16 12:50:34 curl: (28) Operation timed out after 60001 milliseconds with 0 bytes received
```

- Flaky under TSan - #38106

>[Stateless tests (release, s3 storage, actions) — Tests are not finished, fail: 1, passed: 4146, skipped: 52](https://s3.amazonaws.com/clickhouse-test-reports/37469/0d4f78639e94f433e98a580d2c97a61a4193e962/stateless_tests__release__s3_storage__actions_.html)

```
$ pigz -cd clickhouse-server.log.gz | fgrep -a -e 02293_grouping_function_group_by -e 749e5b8a-8567-4c5b-8b08-8452e7d5c48a
...
2022.06.16 16:56:47.786592 [ 929 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Debug> executeQuery: (from [::ffff:127.0.0.1]:43864, initial_query_id: 06c8328c-d06d-417f-b2b0-f46cd50852d2) (comment: 02293_grouping_function_group_by.sql) SELECT `number`, grouping(`number`, `number` % 2) AS `gr` FROM numbers(10) GROUP BY `number`, `number` % 2 WITH ROLLUP WITH TOTALS ORDER BY `number` ASC, `gr` ASC (stage: WithMergeableState)
2022.06.16 16:56:47.786702 [ 929 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> ContextAccess (default): Access granted: CREATE TEMPORARY TABLE ON *.*
2022.06.16 16:56:47.786886 [ 929 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> InterpreterSelectQuery: FetchColumns -> WithMergeableState
2022.06.16 16:56:47.787244 [ 1958 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> AggregatingTransform: Aggregating
2022.06.16 16:56:47.787260 [ 1958 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> Aggregator: An entry for key=15678497920870529023 found in cache: sum_of_sizes=10, median_size=10
2022.06.16 16:56:47.787293 [ 1958 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> Aggregator: Aggregation method: keys128
2022.06.16 16:56:47.787310 [ 1958 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Debug> AggregatingTransform: Aggregated. 10 to 10 rows (from 90.00 B) in 0.000360196 sec. (27762.663 rows/sec., 244.01 KiB/sec.)
2022.06.16 16:56:47.787315 [ 1958 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Trace> Aggregator: Merging aggregated data
2022.06.16 16:56:47.787531 [ 929 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Information> executeQuery: Read 10 rows, 80.00 B in 0.00090523 sec., 11046 rows/sec., 86.30 KiB/sec.
2022.06.16 16:56:47.787573 [ 929 ] {749e5b8a-8567-4c5b-8b08-8452e7d5c48a} <Debug> MemoryTracker: Peak memory usage (for query): 1.06 MiB.
...
2022.06.16 17:06:46.272537 [ 41897 ] {8f4979f8-dc0c-4bb9-aae0-bcdf2224bc95} <Debug> executeQuery: (from [::1]:45314) (comment: 02293_grouping_function_group_by.sql) DROP DATABASE test_ya7omt  (stage: Complete)
...
pigz: skipping: clickhouse-server.log.gz: corrupted -- incomplete deflate data
pigz: abort: internal threads error
```

And `trace_log` is empty, because it is S3.

**Review by KochetovNicolai — APPROVED** [review]
